### PR TITLE
serial/uart/h5: add bt h5 uart serial driver

### PR DESCRIPTION
**Pull request by chengkai15**
## Summary
add bth5 driver

## Impact
none

## Testing
test pass with Realtek Bluetooth (rtl8723)

### CODE REVIEW

File: drivers/serial/uart_bth5.c [hunk 1]
```diff
@@ -0,0 +1,1278 @@
...
+    0x3f, 0xbf, 0x7f, 0xff
+  };
+
+  return rev8table[byte];
+}
+
+static uint16_t
+h5_crc_rev16(uint16_t x)
+{
+  return(h5_crc_rev8(x & 0xff) << 8) | h5_crc_rev8(x >> 8);
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:250]:
```suggestion
  return (h5_crc_rev8(x & 0xff) << 8) | h5_crc_rev8(x >> 8);
```

File: drivers/serial/uart_bth5.c [hunk 2]
```diff
@@ -0,0 +1,1278 @@
...
+}
+
+static void
+h5_crc_update(FAR uint16_t *crc, uint8_t d)
+{
+  uint16_t reg;
+  static const uint16_t crctable[16] =
+  { 0x0000, 0x1081, 0x2102, 0x3183, 0x4204, 0x5285, 0x6306, 0x7387,
+    0x8408, 0x9489, 0xa50a, 0xb58b, 0xc60c, 0xd68d, 0xe70e, 0xf78f
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:259]:
```suggestion
  {
    0x0000, 0x1081, 0x2102, 0x3183, 0x4204, 0x5285, 0x6306, 0x7387,
    0x8408, 0x9489, 0xa50a, 0xb58b, 0xc60c, 0xd68d, 0xe70e, 0xf78f
```

File: drivers/serial/uart_bth5.c [hunk 3]
```diff
@@ -0,0 +1,1278 @@
...
+        ret = 1;
+        break;
+    }
+
+  return ret;
+}
+
+static int
+h5_uart_header(FAR uint8_t *data, enum bt_buf_type_e *type, size_t *pktlen,
+                size_t *hdrlen, size_t reserved)
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:716]:
```suggestion
h5_uart_header(FAR uint8_t *data, FAR enum bt_buf_type_e *type,
               FAR size_t *pktlen, FAR size_t *hdrlen, size_t reserved)
```

File: drivers/serial/uart_bth5.c [hunk 4]
```diff
@@ -0,0 +1,1278 @@
...
+{
+  int ret = OK;
+  FAR union bt_hdr_u *hdr = (FAR union bt_hdr_u *)data;
+
+  switch (*(data - reserved))
+    {
+      case H4_CMD:
+        {
+          *hdrlen = sizeof (struct bt_hci_cmd_hdr_s);
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:725]:
```suggestion
          *hdrlen = sizeof(struct bt_hci_cmd_hdr_s);
```

File: drivers/serial/uart_bth5.c [hunk 5]
```diff
@@ -0,0 +1,1278 @@
...
+          *pktlen = hdr->cmd.param_len;
+          *type = HCI_COMMAND_PKT;
+          break;
+        }
+
+      case H4_ACL:
+        {
+          *hdrlen = sizeof (struct bt_hci_acl_hdr_s);
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:733]:
```suggestion
          *hdrlen = sizeof(struct bt_hci_acl_hdr_s);
```

File: drivers/serial/uart_bth5.c [hunk 6]
```diff
@@ -0,0 +1,1278 @@
...
+          *pktlen = hdr->acl.len;
+          *type = HCI_ACLDATA_PKT;
+          break;
+        }
+
+      case H4_ISO:
+        {
+          *hdrlen = sizeof (struct bt_hci_iso_hdr_s);
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:741]:
```suggestion
          *hdrlen = sizeof(struct bt_hci_iso_hdr_s);
```

File: drivers/serial/uart_bth5.c [hunk 7]
```diff
@@ -0,0 +1,1278 @@
...
+      length += h5_slip_one_byte(frame, length,
+                                 (uint8_t)(h5_txmsg_crc & 0x00ff));
+    }
+
+  length += h5_slip_delim(frame, length);
+
+  work_cancel(HPWORK, &dev->ackworker);
+
+  wlinfo("tx t:%d l:%d s:%d a:%d\n", H5_HDR_PKT_TYPE(hdr), H5_HDR_LEN(hdr),
+          H5_HDR_SEQ(hdr), H5_HDR_ACK(hdr));
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:1055]:
```suggestion
  wlinfo("tx t:%d l:%d s:%d a:%d\n", H5_HDR_PKT_TYPE(hdr), H5_HDR_LEN(hdr),
         H5_HDR_SEQ(hdr), H5_HDR_ACK(hdr));
```

File: drivers/serial/uart_bth5.c [hunk 8]
```diff
@@ -0,0 +1,1279 @@
...
+#define H5_HDR_ACK(hdr)      (((hdr)[0] >> 3) & 0x07)
+#define H5_HDR_CRC(hdr)      (((hdr)[0] >> 6) & 0x01)
+#define H5_HDR_RELIABLE(hdr) (((hdr)[0] >> 7) & 0x01)
+#define H5_HDR_PKT_TYPE(hdr) ((hdr)[1] & 0x0f)
+#define H5_HDR_LEN(hdr)      ((((hdr)[1] >> 4) & 0x0f) + ((hdr)[2] << 4))
+
+#define H5_SET_SEQ(hdr, seq)    ((hdr)[0] |= (seq))
+#define H5_SET_ACK(hdr, ack)    ((hdr)[0] |= (ack) << 3)
+#define H5_SET_RELIABLE(hdr)    ((hdr)[0] |= 1 << 7)
+#define H5_SET_TYPE(hdr, type)  ((hdr)[1] |= type)
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:73]:
```suggestion
#define H5_SET_TYPE(hdr, type)  ((hdr)[1] |= (type))
```

File: drivers/serial/uart_bth5.c [hunk 9]
```diff
@@ -0,0 +1,1279 @@
...
+        {
+          dev->txwin = (data[2] & 0x07);
+          dev->crcvalid = ((data[2] & 0x10) == 0x10);
+          wlinfo("h5 txwin:%d, crcvalid:%d", dev->txwin, dev->crcvalid);
+        }
+
+      if (dev->txwin < CONFIG_UART_BTH5_TXWIN)
+        {
+          wlerr ("h5, txwin(%d) overflow(%d)", dev->txwin,
+                 CONFIG_UART_BTH5_TXWIN);
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:392]:
```suggestion
          wlerr("h5, txwin(%d) overflow(%d)", dev->txwin,
                CONFIG_UART_BTH5_TXWIN);
```

File: drivers/serial/uart_bth5.c [hunk 10]
```diff
@@ -0,0 +1,1279 @@
...
+
+static int
+uart_bth5_open(FAR struct file *filep)
+{
+  FAR struct inode *inode = filep->f_inode;
+  FAR struct uart_bth5_s *dev = inode->i_private;
+  int ret;
+  const uint8_t sync_req[] = {
+    0x01, 0x7e
+  };
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:867]:
```suggestion
  const uint8_t sync_req[] =
  {
    0x01, 0x7e
  };
```

File: drivers/serial/uart_bth5.c [hunk 11]
```diff
@@ -0,0 +1,1279 @@
...
+        {
+          h5_crc_update(&h5_txmsg_crc, payload[idx]);
+        }
+    }
+
+  if (dev->crcvalid)
+    {
+      h5_txmsg_crc = h5_crc_rev16(h5_txmsg_crc);
+      length += h5_slip_one_byte (frame, length,
+                                 (uint8_t)((h5_txmsg_crc >> 8) & 0x00ff));
```
Comment (pkarashchenko) [inline drivers/serial/uart_bth5.c:1046]:
```suggestion
      length += h5_slip_one_byte(frame, length,
                                 (uint8_t)((h5_txmsg_crc >> 8) & 0x00ff));
```

**Review by pkarashchenko — COMMENTED** [review]
Few minor items. Looks good in general

File: drivers/serial/Kconfig [hunk 12]
```diff
@@ -747,4 +747,43 @@ config UART_BTH4_NPOLLWAITERS
 
 endif # UART_BTH4
 
+menuconfig UART_BTH5
+	bool "BT H5 uart pseudo device"
```
Comment (acassis) [inline drivers/serial/Kconfig:751]:
```suggestion
	bool "BT H5 UART Pseudo Device"

File: drivers/serial/Kconfig [hunk 13]
```diff
@@ -747,4 +747,43 @@ config UART_BTH4_NPOLLWAITERS
...
+	---help---
+		Enable support for Bluetooth H5 UART Pseudo Device(eg. /dev/ttyHCI).
+		This instantiates a serial-like interface over an existing bluetooth
+		controller via HCI interface. Useful for external Bluetooth
+		stacks working this way instead of the socket based interface.
+
+if UART_BTH5
+
+config UART_BTH5_TXBUFSIZE
+	int "BT H5 uart TX buffer size"
```
Comment (acassis) [inline drivers/serial/Kconfig:763]:
```suggestion
	int "BT H5 UART TX Buffer Size"

File: drivers/serial/Kconfig [hunk 14]
```diff
@@ -747,4 +747,43 @@ config UART_BTH4_NPOLLWAITERS
...
+	default 2048
+	---help---
+		H5 UART TX buffer size.  Default: 2048
+
+config UART_BTH5_TXWIN
+	int "BT H5 uart TX window size"
```
Comment (acassis) [inline drivers/serial/Kconfig:769]:
```suggestion
	int "BT H5 UART TX Window Size"

File: drivers/serial/Kconfig [hunk 15]
```diff
@@ -747,4 +747,43 @@ config UART_BTH4_NPOLLWAITERS
...
+	default 4
+	---help---
+		H5 UART TX window size.  Default: 4
+
+config UART_BTH5_RXBUFSIZE
+	int "BT H5 uart RX buffer size"
```
Comment (acassis) [inline drivers/serial/Kconfig:775]:
```suggestion
	int "BT H5 UART RX Buffer Size"

File: drivers/serial/Kconfig (hunk 12 above)
Comment (xiaoxiang781216) [reply drivers/serial/Kconfig:751]: @chengkai15 please fix the remain comment.

File: drivers/serial/uart_bth5.c (hunk 8 above)
Comment (xiaoxiang781216) [reply drivers/serial/uart_bth5.c:73]: @chengkai15 fix this too.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**